plugins/utils/spectre.nix

```diff
@@ -0,0 +1,236 @@
+                        description = "The option icon.";
+                      };
+
+                      desc = helpers.mkNullOrStr ''
+                        The description for this option.
+                      '';
+                    };
+                  })
+                  "{}"
+                  "The options for this engine.";
+              };
+            })
+        )
+        ''
+          Definition of the ${type} engines.
+
+          default: see [here](https://github.com/nvim-pack/nvim-spectre/blob/master/lua/spectre/config.lua)
+        '';
+    in {
+      color_devicons = helpers.defaultNullOpts.mkBool true ''
+        Whether to enable color devicons.
+      '';
+
+      open_cmd = helpers.defaultNullOpts.mkStr "vnew" ''
+        The open command.
+      '';
+
+      live_update = helpers.defaultNullOpts.mkBool false ''
+        Auto execute search again when you write to any file in vim.
+      '';
+
+      lnum_for_results = helpers.defaultNullOpts.mkBool false ''
+        Show line number for search/replace results.
+      '';
+
+      line_sep_start =
+        helpers.defaultNullOpts.mkStr
+        "┌──────────────────────────────────────────────────────"
+        "Start of the line separator";
+
+      result_padding = helpers.defaultNullOpts.mkStr "│  " ''
+        Result padding string.
+      '';
+
+      line_sep =
+        helpers.defaultNullOpts.mkStr
+        "└──────────────────────────────────────────────────────"
+        "Line separator.";
+
+      highlight =
+        helpers.defaultNullOpts.mkAttrsOf types.str
+        ''
+          {
+            headers = "SpectreHeader";
+            ui = "SpectreBody";
+            filename = "SpectreFile";
+            filedirectory = "SpectreDir";
+            search = "SpectreSearch";
+            border = "SpectreBorder";
+            replace = "SpectreReplace";
+          }
+        ''
+        "Highlight groups.";
+
+      mapping =
+        helpers.mkNullOrOption
+        (
+          with types;
+            attrsOf (submodule {
+              options = {
+                map = mkOption {
+                  type = types.str;
+                  description = "Keyboard shortcut.";
+                };
+
+                cmd = mkOption {
+                  type = types.str;
+                  description = "Command to run.";
+                  example = "<cmd>lua require('spectre').tab()<cr>";
+                };
+
+                desc = helpers.mkNullOrStr ''
+                  Description for this mapping.
+                '';
+              };
+            })
+        )
+        ''
+          Keymaps declaration.
+
+          default: see [here](https://github.com/nvim-pack/nvim-spectre/blob/master/lua/spectre/config.lua)
+        '';
+
+      find_engine = mkEngineOption "find";
+
+      replace_engine = mkEngineOption "replace";
+
+      default = {
+        find = {
+          cmd = helpers.defaultNullOpts.mkStr "rg" ''
```

Comment:
@traxys do you think that we should automatically install the `find` and `replace` engine ?
```nix
extraPackages = (optional (elem cfg.settings.default.find.cmd ["rg" "ag" ...]) pkgs.${cfg.settings.default.find.cmd}) ++ (optional (elem cfg.settings.default.replace.cmd ["rg" "ag" ...]) pkgs.${cfg.settings.default.find.cmd})
```

Reply:
It would be nicer to do it

Reply:
Is is possible to have the `cmd` option type accept either string or package, and use some `apply` magic to set `cmd` to `lib.getExe pkg`?

If we do an auto-install there should probably be a documented opt-in/out somewhere.

Reply:
> Is is possible to have the `cmd` option type accept either string or package, and use some `apply` magic to set `cmd` to `lib.getExe pkg`?

According to me, this would be against the concept of RFC-42 `settings` and pretty unusual (at least from Nixvim). But why not ?


> If we do an auto-install there should probably be a documented opt-in/out somewhere.
Yes, actually I wanted to go without any installation at all here. @traxys what did you have in mind exactly ?

Reply:
I feel like there is no good option here, that satisfies:
- automatically detect the values of `settings.default.find.cmd`
- Allow the user to prevent the auto-install or change the package
- Conform with RFC 42

Hence, I would advocate to ship this plugin without any installation.
If the user wants to have `rg` installed, he can add it to `extraPackages`.
Also, we could simply add a warning in the description.